### PR TITLE
CEL: Various improvements

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by sbueringer** [comment]
cc @jimmidyson

File: api/v1beta1/clusterclass_types.go [hunk 1]
```diff
@@ -454,6 +454,16 @@ type JSONSchemaProps struct {
...
+	// MaxProperties is the maximum amount of entries in a map or properties in an object.
+	// NOTE: Can only be set if type is object.
+	// +optional
+	MaxProperties *int64 `json:"maxProperties,omitempty"`
+
+	// MinProperties is the minimum amount of entries in a map or properties in an object.
+	// NOTE: Can only be set if type is object.
+	// +optional
+	MinProperties *int64 `json:"minProperties,omitempty"`
+
```
Comment (jimmidyson) [inline api/v1beta1/clusterclass_types.go:466]: I like having this for variable schemas, but should this be in a separate PR?

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 2]
```diff
@@ -1619,22 +1930,127 @@ func Test_ValidateClusterVariable(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "Error if integer is below minimum via CEL expression with custom error message (messageExpression preferred)",
+			wantErrs: []validationMatch{
+				invalid("Invalid value: \"0\": new value must be greater than or equal to 1, but got 0",
```
Comment (jimmidyson) [inline internal/topology/variables/cluster_variable_validation_test.go:1936]: I like the addition of the value in here! Thank you for this!

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 3]
```diff
@@ -156,43 +190,82 @@ func validateRootSchema(ctx context.Context, clusterClassVariable *clusterv1.Clu
...
+	//   * If a ClusterClass update keeps a CEL expression unchanged, it will be validated with the "max" env
+	//   * On Cluster create or update, variable values will always be validated with the "max" env
+	// * The goal is to only allow adding new CEL expressions that are compatible with "n-1",
+	//   while allowing to use pre-existing CEL expressions that are compatible with "max".
+	// * We have to do this because:
+	//   * This makes it possible to rollback Cluster API from "m" to "m-1", because Cluster API "m-1" is
+	//     able to use CEL expressions created with Cluster API "m" (because "m" only allows to add CEL expressions
+	//     that can be used by "m-1").
+	//   * The Kubernetes CEL library assumes this behavior and it's baked into cel.NewValidator (they use n to validate)
+	// * If the DefaultCompatibilityVersion is an earlier version than "n-1", it means we could roll back even more than 1 version.
```
Comment (jimmidyson) [inline internal/topology/variables/clusterclass_variable_validation.go:242]: Great explanation!

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 4]
```diff
@@ -217,166 +290,218 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path,
...
-		} else if rule.Message != "" && trimmedMsg == "" {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must be non-empty if specified"))
-		} else if rule.Message != "" && len(rule.Message) > 2048 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must have a maximum length of 2048 characters"))
-		} else if hasNewlines(trimmedMsg) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must not contain line breaks"))
-		} else if hasNewlines(trimmedRule) && trimmedMsg == "" {
-			allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), "message must be specified if rule contains line breaks"))
+	if len(schema.XValidations) > 0 {
+		// Return if schema is not a structural schema (this should enver happen, but let's handle it anyway).
```
Comment (jimmidyson) [inline internal/topology/variables/clusterclass_variable_validation.go:333]: nit

```suggestion
		// Return if schema is not a structural schema (this should never happen, but let's handle it anyway).
```

File: api/v1beta1/clusterclass_types.go (hunk 1 above)
Comment (sbueringer) [reply api/v1beta1/clusterclass_types.go:466]: Just surfacing here from DM, I couldn't make the cost thing work correctly without it
The problem is that properties and additionalProperties are unbound without it
